### ISSUE TITLE
After upgrading to Node 18, yarn cli fails to run postinstall script when yarn cache is enabled

### ISSUE THREAD

**Issue by henhal** [actions/setup-node#902]
**Description:**
After updating to Node18 + serverless v3, my GHA builds fail consistency after first succeeding once (when there is a cache miss).
The issue appears to be related to the yarn cache, but only happens on GHA.
Disabling the yarn cache resolves the issue.

**Action version:**
actions/setup-node@v3

**Platform:**
- [X] Ubuntu
- [ ] macOS
- [ ] Windows

**Runner type:**
- [X] Hosted
- [ ] Self-hosted

**Tools version:**

node 18.18.2
yarn 1.22.21

**Repro steps:**  

Set up a repo with 

```
"dependencies": {
    "serverless": "^3.37.0",
    "serverless-bundle": "^6.0.0"
}
````

* Running `yarn install` on my local machine works fine.
* The first build on GHA (in the pull request) worked fine
* Consecutive builds on GHA fails 100%.
* Removing `cache: yarn` from the action solves the issue (but obviously adds build time)

Error message seems slightly random, but in my case always related to a postinstall script in `serverless`:

```
[4/4] Building fresh packages...
info This package requires node-gyp, which is not currently installed. Yarn will attempt to automatically install it. If this fails, you can run "yarn global add node-gyp" to manually install it.
error TypeError: /home/runner/work/<mymodule>/node_modules/serverless: Cannot read properties of undefined (reading 'config')
info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.
    at /usr/local/lib/node_modules/yarn/lib/cli.js:[34](https://github.com/<mymodule>/actions/runs/6955033577/job/18925299805#step:5:35)1[35](https://github.com/<mymodule>/actions/runs/6955033577/job/18925299805#step:5:36):120
    at Generator.next (<anonymous>)
    at step (/usr/local/lib/node_modules/yarn/lib/cli.js:310:30)
    at /usr/local/lib/node_modules/yarn/lib/cli.js:321:13
```

Note that prior to upgrading to node 18 and serverless 3, it worked fine on GHA as well, using Node 16.20.2 and the following dependencies:

package.json:
```
"dependencies": {
    "serverless": "^2.72.3",
    "serverless-bundle": "^5.3.0"
}
```

I added `--verbose` to my `yarn install --frozen-lockfile` command, but nothing useful came out.

**Expected behavior:**

`yarn install --frozen-lockfile` should work also with yarn cache enabled

**Actual behavior:**
`yarn install --frozen-lockfile` fails on GHA with Node 18.x when yarn cache is enabled.

**Comment by nikolai-laevskii** [comment]
Thank you for creating the issue and providing us with details! We will make sure to look into it and come back with the answer!

**Comment by henhal** [comment]
Today I've experienced this bug also with cache disabled. :( Really strange, as I said, the first time GHA installed my project it worked fine. The second, third etc, failed. After disabling the cache, it worked a couple of times. Now it fails again.

The line in `yarn` that throws is typically

```
var keys = new (_set || _load_set()).default([].concat((0, (_keys || _load_keys()).default)(config.registries.yarn.config), (0, (_keys || _load_keys()).default)(config.registries.npm.config)));
```

due to `config.registries.yarn` being `undefined`.

**Comment by wouterds** [comment]
Can also reproduce fwiw

**Comment by henhal** [comment]
I had no choice but to temporarily switch from yarn to npm as this resolves the problems. So it only happens with yarn on setup/node action. Not with yarn locally.

**Comment by henhal** [comment]
I haven't been able to find any other workaround. :( 
Luckily it was not a super big deal for me to switch from yarn to npm, other than getting used to writing the correct command after using yarn for years :(
But since it seems to be reproducible I'm hoping the maintainers at least have a good chance of debugging this. I am not able to share my full configuration as the code in question is not public, so was a bit worried that it was due to some very specific quirk in my setup, but now it definitely seems like others are seeing a very similar error.

**Comment by wouterds** [comment]
I could work around it by adding the ` --ignore-scripts` flag to my yarn install commands in GHA workflows.

**Comment by priya-kinthali** [comment]
Hello @henhal , Thank you for reporting the issue.
We tried to reproduce the issue in our local and the error message mentioned in description appears to come from the installation process of the serverless package.Could you please try reinstalling the packages and have node-gyp globally installed as suggested in the error message.
In case of any concerns please feel free to share them!!

**Comment by aparnajyothi-y** [comment]
Hello @henhal,
Just a gentle reminder, Could you please try reinstalling the packages and have node-gyp globally installed as suggested in the error message.

**Comment by wouterds** [comment]
Stopped using yarn, not a great workaround but oh well

**Comment by tiborsaas-tw** [comment]
Anyone running into this, try installing Node 20+. In our case it solved this exact same issue with `node-gyp`.

**Comment by kddsultan** [comment]
> Hello @henhal , Thank you for reporting the issue. We tried to reproduce the issue in our local and the error message mentioned in description appears to come from the installation process of the serverless package.Could you please try reinstalling the packages and have node-gyp globally installed as suggested in the error message. In case of any concerns please feel free to share them!!

It would be useful to know why this is happening in versions starting with Node v18.18, but got resolved in v20+.

**Comment by vinaygregoryjohn183** [comment]
@kddsultan I'm also facing the issue you're having wherein the node version is getting resolved to v20 even when i added version >=18.0.0 in the engines section in package.json
